File: results/0.25/Bitcoin-transactions/bitcoin-transactions-claude-3-5-haiku-20241022.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <hammer/hammer.h>
#include <hammer/allocator.h>

typedef struct {
    uint32_t version;
    size_t input_count;
    HParser* inputs;
    size_t output_count;
    HParser* outputs;
    uint32_t locktime;
} BitcoinTransaction;

typedef struct {
    uint8_t prev_tx_hash[32];
    uint32_t prev_output_index;
    size_t scriptsig_length;
    uint8_t* scriptsig;
    uint32_t sequence_number;
} TransactionInput;

typedef struct {
    uint64_t amount;
    size_t scriptpubkey_length;
    uint8_t* scriptpubkey;
} TransactionOutput;

HParser* bitcoin_transaction_parser() {
    HArena* arena = h_new_arena();
    
    HParser* version = h_int32();
    
    HParser* var_int = h_choice(
        h_uint8(),
        h_uint16(),
        h_uint32(),
        h_uint64(),
        NULL
    );
    
    HParser* input_count_parser = var_int;
    
    HParser* input_parser = h_sequence(
        h_repeat_n(h_uint8(), 32),  // prev tx hash
        h_uint32(),                 // prev output index
        var_int,                    // scriptsig length
        h_length_value(h_uint8(), var_int),  // scriptsig
        h_uint32(),                 // sequence number
        NULL
    );
    
    HParser* inputs_parser = h_many(input_parser);
    
    HParser* output_count_parser = var_int;
    
    HParser* output_parser = h_sequence(
        h_uint64(),                 // amount
        var_int,                    // scriptpubkey length
        h_length_value(h_uint8(), var_int),  // scriptpubkey
        NULL
    );
    
    HParser* outputs_parser = h_many(output_parser);
    
    HParser* locktime = h_uint32();
    
    HParser* transaction_parser = h_sequence(
        version,
        input_count_parser,
        inputs_parser,
        output_count_parser,
        outputs_parser,
        locktime,
        NULL
    );
    
    return transaction_parser;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        fprintf(stderr, "Usage: %s <bitcoin_transaction_file>\n", argv[0]);
        return 1;
    }
    
    FILE* file = fopen(argv[1], "rb");
    if (!file) {
        perror("Error opening file");
        return 1;
    }
    
    fseek(file, 0, SEEK_END);
    long file_size = ftell(file);
    rewind(file);
    
    uint8_t* buffer = malloc(file_size);
    if (!buffer) {
        perror("Memory allocation error");
        fclose(file);
        return 1;
    }
    
    size_t read_size = fread(buffer, 1, file_size, file);
    fclose(file);
    
    if (read_size != file_size) {
        perror("File read error");
        free(buffer);
        return 1;
    }
    
    HArena* arena = h_new_arena();
    HParser* parser = bitcoin_transaction_parser();
    HParseResult* result = h_parse(parser, buffer, file_size);
    
    if (result && result->ast) {
        printf("Transaction parsed successfully\n");
    } else {
        printf("Transaction parsing failed\n");
    }
    
    h_parse_result_free(result);
    h_arena_free(arena, parser);
    h_arena_free(arena, buffer);
    
    return 0;
}

File: results/0.25/Bitcoin-transactions/bitcoin-transactions-claude-3-5-haiku-20241022.rs
use nom::{
    bytes::complete::take,
    combinator::map,
    multi::length_count,
    number::complete::{le_u16, le_u32, le_u64, le_u8},
    IResult,
};
use std::env;
use std::fs::File;
use std::io::Read;

#[derive(Debug)]
struct TransactionInput {
    prev_tx_hash: [u8; 32],
    prev_output_index: u32,
    script_sig_length: u64,
    script_sig: Vec<u8>,
    sequence: u32,
}

#[derive(Debug)]
struct TransactionOutput {
    amount: u64,
    script_pubkey_length: u64,
    script_pubkey: Vec<u8>,
}

#[derive(Debug)]
struct Transaction {
    version: u32,
    inputs: Vec<TransactionInput>,
    outputs: Vec<TransactionOutput>,
    locktime: u32,
}

fn parse_varint(input: &[u8]) -> IResult<&[u8], u64> {
    let (input, first_byte) = le_u8(input)?;
    match first_byte {
        0..=0xFC => Ok((input, first_byte as u64)),
        0xFD => map(le_u16, |x| x as u64)(input),
        0xFE => map(le_u32, |x| x as u64)(input),
        0xFF => le_u64(input),
    }
}

fn parse_transaction_input(input: &[u8]) -> IResult<&[u8], TransactionInput> {
    let (input, prev_tx_hash) = take(32usize)(input)?;
    let (input, prev_output_index) = le_u32(input)?;
    let (input, script_sig_length) = parse_varint(input)?;
    let (input, script_sig) = take(script_sig_length as usize)(input)?;
    let (input, sequence) = le_u32(input)?;

    Ok((input, TransactionInput {
        prev_tx_hash: prev_tx_hash.try_into().unwrap(),
        prev_output_index,
        script_sig_length,
        script_sig: script_sig.to_vec(),
        sequence,
    }))
}

fn parse_transaction_output(input: &[u8]) -> IResult<&[u8], TransactionOutput> {
    let (input, amount) = le_u64(input)?;
    let (input, script_pubkey_length) = parse_varint(input)?;
    let (input, script_pubkey) = take(script_pubkey_length as usize)(input)?;

    Ok((input, TransactionOutput {
        amount,
        script_pubkey_length,
        script_pubkey: script_pubkey.to_vec(),
    }))
}

fn parse_transaction(input: &[u8]) -> IResult<&[u8], Transaction> {
    let (input, version) = le_u32(input)?;
    let (input, inputs) = length_count(parse_varint, parse_transaction_input)(input)?;
    let (input, outputs) = length_count(parse_varint, parse_transaction_output)(input)?;
    let (input, locktime) = le_u32(input)?;

    Ok((input, Transaction {
        version,
        inputs,
        outputs,
        locktime,
    }))
}

fn main() -> std::io::Result<()> {
    let args: Vec<String> = env::args().collect();
    if args.len() != 2 {
        eprintln!("Usage: {} <transaction_file>", args[0]);
        std::process::exit(1);
    }

    let mut file = File::open(&args[1])?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)?;

    match parse_transaction(&buffer) {
        Ok((_, transaction)) => {
            println!("Parsed Transaction: {:?}", transaction);
            Ok(())
        }
        Err(e) => {
            eprintln!("Failed to parse transaction: {:?}", e);
            std::process::exit(1)
        }
    }
}